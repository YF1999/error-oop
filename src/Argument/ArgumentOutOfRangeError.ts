import { IErrorArguments } from '../Native';
import { ErrorTool } from '../Tools';
import { ArgumentError, IArgumentErrorOptions } from './ArgumentError';

export interface IArgumentOutOfRangeErrorOptions extends IArgumentErrorOptions {
    /**
     * The value of the argument that causes this error.
     */
    actualValue?: unknown;
}

/**
 * Applicable when the value of an argument is outside the allowable range of values as defined by the invoked function
 * or method.
 */
export class ArgumentOutOfRangeError extends ArgumentError {
    protected _actualValue?: unknown;

    public constructor();
    /**
     * @param message The error message that explains the reason for this error.
     */
    public constructor(message: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param innerError The error that is the cause of the current error. Stack trace will be appended.
     */
    public constructor(message: string, innerError: Error);
    /**
     * @param message The error message that explains the reason for this error.
     * @param paramName The name of the parameter that caused the current error.
     */
    public constructor(message: string, paramName: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param paramName The name of the parameter that caused the current error.
     * @param innerError The error that is the cause of the current error. Stack trace will be appended.
     */
    public constructor(message: string, paramName: string, innerError: Error);
    /**
     * @param message The error message that explains the reason for this error.
     * @param paramName The name of the parameter that caused the current error.
     * @param actualValue The value of the argument that causes this error.
     */
    public constructor(message: string, paramName: string, actualValue: unknown);
    /**
     * @param message The error message that explains the reason for this error.
     * @param paramName The name of the parameter that caused the current error.
     * @param actualValue The value of the argument that causes this error.
     * @param innerError The error that is the cause of the current error. Stack trace will be appended.
     */
    public constructor(message: string, paramName: string, actualValue: unknown, innerError: Error);
    /**
     * @param options The constructor options.
     */
    public constructor(options: IArgumentOutOfRangeErrorOptions);

    public constructor(...args: IErrorArguments<IArgumentOutOfRangeErrorOptions, ['paramName', 'actualValue']>) {
        const opts = ErrorTool.parseErrorArguments(['paramName', 'actualValue'], ...args);

        if ('actualValue' in opts && opts.paramName && opts.noMessageSuffix !== true) {
            opts.noMessageSuffix = true;
            opts.message = `${opts.message} (Parameter '${opts.paramName}', ActualValue '${opts.actualValue}')`;
        }

        super(opts);

        this._actualValue = opts.actualValue;
        this._setNonEnumerable('_actualValue');
    }

    public get actualValue(): unknown | undefined {
        return this._actualValue;
    }
}
