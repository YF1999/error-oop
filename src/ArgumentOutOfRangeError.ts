import { ErrorOptions, ArgumentOutOfRangeMessageProps, ArgumentOutOfRangeErrorProps } from './CommonTypes';
import { AbstractArgumentError } from './ArgumentError';

export abstract class AbstractArgumentOutOfRangeError extends AbstractArgumentError {
    protected _actualValue?: unknown;

    public constructor(props: ArgumentOutOfRangeErrorProps, options: ErrorOptions<ArgumentOutOfRangeMessageProps>) {
        const { actualValue } = props;
        const { generateMessage: gm, ...others } = options;

        super(props, {
            generateMessage: gm && ((_props) => gm({ ..._props, actualValue })),
            ...others,
        });

        this._actualValue = actualValue;
        this._setNonEnumerable('_actualValue');
    }

    public get actualValue(): unknown | undefined {
        return this._actualValue;
    }
}

/**
 * Applicable when the value of an argument is outside the allowable range of values as defined by
 * the invoked function or method.
 */
export class ArgumentOutOfRangeError extends AbstractArgumentOutOfRangeError {
    public constructor();
    /**
     * @param message The error message that explains the reason for this error.
     */
    public constructor(message: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param innerError The error that is the cause of the current error. Stack trace will be append.
     */
    public constructor(message: string, innerError: Error);
    /**
     * @param message The error message that explains the reason for this error.
     * @param paramName The name of the parameter that caused the current error.
     */
    public constructor(message: string, paramName: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param paramName The name of the parameter that caused the current error.
     * @param innerError The error that is the cause of the current error. Stack trace will be append.

     */
    public constructor(message: string, paramName: string, innerError: Error);
    /**
     * @param message The error message that explains the reason for this error.
     * @param paramName The name of the parameter that caused the current error.
     * @param actualValue The value of the argument that causes this error.
     */
    public constructor(message: string, paramName: string, actualValue: unknown);
    /**
     * @param message The error message that explains the reason for this error.
     * @param paramName The name of the parameter that caused the current error.
     * @param actualValue The value of the argument that causes this error.
     * @param innerError The error that is the cause of the current error. Stack trace will be append.
     */
    public constructor(message: string, paramName: string, actualValue: unknown, innerError: Error);

    public constructor(message: string = '', arg1?: string | Error, arg2?: unknown, arg3?: Error) {
        function generateMessage(props: ArgumentOutOfRangeMessageProps) {
            let append = '';
            if (props.paramName && props.actualValue) {
                append = ` (Parameter '${props.paramName}', ActualValue '${props.actualValue}')`;
            } else if (props.paramName && !props.actualValue) {
                append = ` (Parameter '${props.paramName}')`;
            }

            return props.message + append;
        }

        // message + innerError?
        if (arg1 === undefined || typeof arg1 !== 'string') {
            super({ message, innerError: arg1 }, { generateMessage });
        }

        // message + paramName + innerError?
        else if (arg2 === undefined || arg2 instanceof Error) {
            super({ message, innerError: arg2, paramName: arg1 }, { generateMessage });
        }

        // message + paramName + actualValue + innerError?
        else {
            super({ message, actualValue: arg2, innerError: arg3, paramName: arg1 }, { generateMessage });
        }
    }
}
