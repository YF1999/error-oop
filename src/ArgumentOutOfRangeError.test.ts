import { ArgumentOutOfRangeError } from './ArgumentOutOfRangeError';

describe('Test `ArgumentOutOfRangeError`', () => {
    const name = 'ArgumentOutOfRangeError';
    const message = 'error';
    const paramName = 'key_parameter';
    const actualValue = 'wrong_value';

    function getErrors() {
        const err1 = new ArgumentOutOfRangeError();
        const err2 = new ArgumentOutOfRangeError(message);
        const err3 = new ArgumentOutOfRangeError(message, err2);
        const err4 = new ArgumentOutOfRangeError(message, paramName, err3);
        const err5 = new ArgumentOutOfRangeError(message, paramName, actualValue, err4);
        const err6 = new ArgumentOutOfRangeError({ message, innerError: err5, paramName });
        const err7 = new ArgumentOutOfRangeError({ message, innerError: err6, noMessageSuffix: true });
        const err8 = new ArgumentOutOfRangeError({
            message,
            innerError: err7,
            paramName,
            actualValue,
        });

        return [err1, err2, err3, err4, err5, err6, err7, err8];
    }

    test('property: error name', () => {
        getErrors().map((err) => expect(err.name).toBe(name));
    });

    test('property: message', () => {
        const [err1, ...errs] = getErrors();
        expect(err1.message).toBe('');
        errs.map((err) => expect(err.message).toMatchSnapshot());
    });

    test('property: inner error', () => {
        const [err1, err2, err3, err4, err5, err6, err7, err8] = getErrors();
        expect(err1.innerError).toBeUndefined();
        expect(err2.innerError).toBeUndefined();
        expect(err3.innerError).toBe(err2);
        expect(err4.innerError).toBe(err3);
        expect(err5.innerError).toBe(err4);
        expect(err6.innerError).toBe(err5);
        expect(err7.innerError).toBe(err6);
        expect(err8.innerError).toBe(err7);
    });

    test('property: parameter name', () => {
        const [err1, err2, err3, err4, err5, err6, err7, err8] = getErrors();
        expect(err1.paramName).toBeUndefined();
        expect(err2.paramName).toBeUndefined();
        expect(err3.paramName).toBeUndefined();
        expect(err4.paramName).toBe(paramName);
        expect(err5.paramName).toBe(paramName);
        expect(err6.paramName).toBe(paramName);
        expect(err7.paramName).toBeUndefined();
        expect(err8.paramName).toBe(paramName);
    });

    test('property: actual value', () => {
        const [err1, err2, err3, err4, err5, err6, err7, err8] = getErrors();
        expect(err1.actualValue).toBeUndefined();
        expect(err2.actualValue).toBeUndefined();
        expect(err3.actualValue).toBeUndefined();
        expect(err4.actualValue).toBeUndefined();
        expect(err5.actualValue).toBe(actualValue);
        expect(err6.actualValue).toBeUndefined();
        expect(err7.actualValue).toBeUndefined();
        expect(err8.actualValue).toBe(actualValue);
    });

    test('empty serialization', () => {
        getErrors().map((err) => expect(JSON.stringify(err)).toBe('{}'));
    });

    test('no enumerable fields', () => {
        getErrors().map((err) => expect(Object.keys(err).length).toBe(0));
    });

    test('own keys', () => {
        const [err1, ...errs] = getErrors();
        expect(Reflect.ownKeys(err1)).toStrictEqual(['stack', '_paramName', '_actualValue']);
        errs.map((err) =>
            expect(Reflect.ownKeys(err)).toStrictEqual(['stack', 'message', '_paramName', '_actualValue']),
        );
    });
});
