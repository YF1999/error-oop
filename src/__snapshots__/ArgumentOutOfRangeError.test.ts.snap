// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test \`ArgumentOutOfRangeError\` property: message 1`] = `"error"`;

exports[`Test \`ArgumentOutOfRangeError\` property: message 2`] = `"error"`;

exports[`Test \`ArgumentOutOfRangeError\` property: message 3`] = `"error (Parameter 'key_parameter')"`;

exports[`Test \`ArgumentOutOfRangeError\` property: message 4`] = `"error (Parameter 'key_parameter', ActualValue 'wrong_value')"`;

exports[`Test \`ArgumentOutOfRangeError\` property: message 5`] = `"error (Parameter 'key_parameter')"`;

exports[`Test \`ArgumentOutOfRangeError\` property: message 6`] = `"error"`;

exports[`Test \`ArgumentOutOfRangeError\` property: message 7`] = `"error (Parameter 'key_parameter', ActualValue 'wrong_value')"`;
